x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/cohere/rerank/CohereRerankServiceSettings.java

```diff
@@ -7,43 +7,115 @@
 
 package org.elasticsearch.xpack.inference.services.cohere.rerank;
 
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 import org.elasticsearch.TransportVersion;
 import org.elasticsearch.TransportVersions;
 import org.elasticsearch.common.ValidationException;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.inference.ModelConfigurations;
 import org.elasticsearch.inference.ServiceSettings;
+import org.elasticsearch.inference.SimilarityMeasure;
 import org.elasticsearch.xcontent.XContentBuilder;
 import org.elasticsearch.xpack.inference.services.ConfigurationParseContext;
-import org.elasticsearch.xpack.inference.services.cohere.CohereServiceSettings;
+import org.elasticsearch.xpack.inference.services.cohere.CohereRateLimitServiceSettings;
+import org.elasticsearch.xpack.inference.services.cohere.CohereService;
 import org.elasticsearch.xpack.inference.services.settings.FilteredXContentObject;
+import org.elasticsearch.xpack.inference.services.settings.RateLimitSettings;
 
 import java.io.IOException;
+import java.net.URI;
 import java.util.Map;
 import java.util.Objects;
 
-public class CohereRerankServiceSettings extends FilteredXContentObject implements ServiceSettings {
+import static org.elasticsearch.xpack.inference.services.ServiceFields.MODEL_ID;
+import static org.elasticsearch.xpack.inference.services.ServiceFields.URL;
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.convertToUri;
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.createOptionalUri;
+import static org.elasticsearch.xpack.inference.services.ServiceUtils.extractOptionalString;
+import static org.elasticsearch.xpack.inference.services.cohere.CohereServiceSettings.DEFAULT_RATE_LIMIT_SETTINGS;
+import static org.elasticsearch.xpack.inference.services.cohere.CohereServiceSettings.OLD_MODEL_ID_FIELD;
+
+public class CohereRerankServiceSettings extends FilteredXContentObject implements ServiceSettings, CohereRateLimitServiceSettings {
     public static final String NAME = "cohere_rerank_service_settings";
 
-    public static CohereRerankServiceSettings fromMap(Map<String, Object> map, ConfigurationParseContext parseContext) {
+    private static final Logger logger = LogManager.getLogger(CohereRerankServiceSettings.class);
+
+    public static CohereRerankServiceSettings fromMap(Map<String, Object> map, ConfigurationParseContext context) {
         ValidationException validationException = new ValidationException();
-        var commonServiceSettings = CohereServiceSettings.fromMap(map, parseContext);
+
+        String url = extractOptionalString(map, URL, ModelConfigurations.SERVICE_SETTINGS, validationException);
```

Comment:
A model config created in an earlier version will still have the `dims`, `max_input_tokens` ... fields persisted. When the config is read back here this class should accept and ignore those settings otherwise I think there will be an error about parsing an unknown setting. The rolling upgrade test `CohereServiceUpgradeIT` should catch such an error if it occurs.

Unfortunately `CohereServiceUpgradeIT` is muted, please unmute it in this PR and add the `test-full-bwc` label, let's see what the test finds.


Unmuting `CohereServiceUpgradeIT` permanently is not part of this PR we should do that separately

Reply:
Thanks for the explanation!

> When the config is read back here this class should accept and ignore those settings otherwise I think there will be an error about parsing an unknown setting.

This happens, because of this line right?
https://github.com/elastic/elasticsearch/blob/ffea002a99554e72bfe0954d838857d523935942/x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/cohere/CohereService.java#L91

I'll unmute the test first without adding accepting `dims` and `max_input_tokens` to check, whether the test finds this error and add it afterwards 👍

Reply:
That's the line. 

In general the code should be lenient when parsing previously persisted configurations and ignore unknown fields. We have many inference services now and they all should have upgrade tests, if we can refactor the parsing logic to make it common between all the services it will be much easier to test. 

You can run the tests locally with
`./gradlew :x-pack:plugin:inference:qa:rolling-upgrade:bwcTest` 
or 
`./gradlew :x-pack:plugin:inference:qa:rolling-upgrade:v8.15.0#bwcTest`